### ISSUE TITLE
Modules.override() does not work with @Provides methods

### ISSUE THREAD

**Issue by pisarek** [google/guice#1029]
Hi,

When I do something like:
```java
public class ModuleA extends AbstractModule {
    @Override
    protected void configure() {
        //some bindings
    }

    @Provides
    @Singleton
    @SomeAnnotation
    Thing createThing(
            <<dependencies>>)) {
        return new Thing(dependencies);
    }
}

public class ModuleB extends AbstractModule {
    @Override
    protected void configure() {
        //some bindings
    }

    @Provides
    @Singleton
    @SomeAnnotation
    Thing createThing(
            <<different dependencies>>)) {
        return new Thing(different dependencies);
    }
}
```

and then do
`Modules.override(new ModuleA()).with(new ModuleB());`
Guice still asks for dependencies from ModuleA during injector creation.
But when I change the methods to Provider classes and bind them like this:
`bind(Thing.class).annotatedWith(SomeAnnotation.class).toProvider(ThingProvider.class).in(Singleton.class);`
Guice does not ask for them.

What is going on?

**Comment by lukesandberg** [comment]
I believe this is a side effect of how `@Provides` methods are implemented, since they always call Binder.getProvider() for every dependency and `Modules.override` doesn't dedupe or remove `ProviderLookup` elements

https://github.com/google/guice/blob/master/core/src/com/google/inject/internal/ProviderMethodsModule.java#L259

As it happens, this bit of code will be removed in the next version of guice and as such `Modules.override` should no longer exhibit this behavior.

**Comment by chriskessel** [comment]
What release will this be fixed in? I just ran into this and it took a few hours of debugging to figure out why.

BTW, switching to @Named seems to work...

**Comment by ajitak** [comment]
@lukesandberg Which guice version has the fix you had mentioned?

**Comment by WrenPriest** [comment]
@ajitak The issue was fixed in [Guice 4.2](https://github.com/google/guice/releases/tag/4.2).
